File: problems/0006_zigzag-conversion/solution.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
    'use strict'
    if (s.length <= 2 || numRows === 1)
        return s
    const colsPerPad = numRows - 1,
          charsPerPad = colsPerPad - 1 + numRows,
          lastRow = colsPerPad
    let result = ''
    for (let r = 0; r < numRows; ++r)
        for (let i = r; i < s.length; i += charsPerPad) {
            result += s[i]
            if (r != 0 && r != lastRow) {
                const j = i + charsPerPad - 2 * r
                if (j < s.length)
                    result += s[j]
            }
        }
    return result
};

File: problems/0006_zigzag-conversion/solution.rb
#!/usr/bin/ruby

# @param {String} s
# @param {Integer} num_rows
# @return {String}
def convert(s, num_rows)
    if (s.length <= 2 || num_rows == 1)
        return s
    end
    cols_per_pad = num_rows - 1
    chars_per_pad = cols_per_pad - 1 + num_rows
    last_row = cols_per_pad
    result = ''
    for r in 0..(num_rows - 1)
        for i in (r..(s.length - 1)).step(chars_per_pad)
            result << s[i]
            j = i + chars_per_pad - 2 * r
            if (r != 0 and r != last_row and j < s.length)
                result << s[j]
            end
        end
    end
    return result
end
